### PR TITLE
Draft PR for VyOS init script is out of date #885 v2

### PR DESCRIPTION
**Pull request by Giggum**
Change of approach vs previous draft PR to use systemd host service with adaptation to save to persitent storage for VyOS.

Without this, the /etc/nextdns.conf and /etc/systemd/system/nextdns.service files are deleted every time VyOS in upgraded

Install.sh also adapted to save binary to persistent storage.

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -1,4 +1,7 @@
-module github.com/nextdns/nextdns
+//module github.com/nextdns/nextdns
+module module github.com/Giggum/nextdns/add-vyos-885_2/
+replace github.com/nextdns/nextdns v0.0.0 => github.com/Giggum/nextdns/add-vyos-885_2 v1.1.1
+
```
Comment (rs) [inline go.mod:4]: Please keep this change local

File: host/service/systemd/service.go [hunk 2]
```diff
@@ -19,10 +19,29 @@ type Service struct {
...
 
+func isVyOS() bool {
+	if st, err := os.Stat("/config/scripts/"); err != nil || !st.IsDir() {
+		if _, err = os.Stat("/usr/libexec/vyos/init/vyos-router"); err != nil {
+			return true
+		}
+	}
+	return false
+}
+
```
Comment (rs) [inline host/service/systemd/service.go:30]: Please move this in a VyOS package which just implement the New function and return a systemd Service with the custom Path/Config. The systemd package must remain free of OS specific code.

Comment (Giggum) [reply host/service/systemd/service.go:30]: Noted

**Comment by Giggum** [comment]
@rs would like to test the progress so far

**Comment by rs** [comment]
Any reason for duplicating the while systemd package?

**Comment by Giggum** [comment]
> Any reason for duplicating the whole systemd package?

Thinking was that since vyos=bookworm use the systemd file as basis. Is there a better suggestion?

**Comment by Giggum** [comment]
> 📷 [Snapshot created](https://snapshot.nextdns.io/pr-921/index.html)

Failed see console output below -> installer should be looking for "nextdns-SNAPSHOT-033b214_linux_amd64.tar.gz" which I see is present. I downloaded and was able to manually untar that package.

Screenshot 2024-04-05 at 9 46 46 AM

**Comment by rs** [comment]
My suggestion was to only implement `New()` in the vyos package and import the `Service` implementation from the `systemd` package.

**Comment by rs** [comment]
To test the install script, you need to point the url to the install script of this PR as well.

**Comment by Giggum** [comment]
> To test the install script, you need to point the url to the install script of this PR as well.

2nd try with sourcing of the submitted install script and same error as before:
Screenshot 2024-04-05 at 1 11 56 PM

**Comment by Giggum** [comment]
> My suggestion was to only implement `New()` in the `vyos` package and import the `Service` implementation from the `systemd` package.

Makes sense, will submit that as a PR once I get the basics going.

The things I want to get right are:
- Get a basic installation working with binary and configuration stored in /config
- Link the .service file in /config/nextdns to /etc/systemd/system/nextdns.service
- Link binary in /config/nextdns to /usr/bin/nextdns (so that it's in PATH)
- Figure out the best way to handle placing 127.0.0.1 in /etc/resolv.conf currently it's through a direct edit during installation but any changes to DNS through vyos will override resolv.conf with whatever configuration is stored in /config/config.boot.. The better approach would be to issue vyos commands:
1) delete system name-server
2) set system name-server '127.0.0.1'
3) commit (to save entry to config.boot)
4) save

**Comment by Giggum** [comment]
> My suggestion was to only implement `New()` in the `vyos` package and import the `Service` implementation from the `systemd` package.

This would mean adding "github.com/nextdns/nextdns/host/service/system" to import and then adapting each service function to call the equivalent systemd.service() correct?

For example the vyos install becomes:
`
func (s Service) Install() error {
	systemd.Install()
}
`

**Comment by rs** [comment]
New would just return a systemd.Service with the right config. Nothing else is needed.

**Comment by rs** [comment]
Please test the implementation in main

**Pull request closed without merging**